prov/gni/src/gnix_mbox_allocator.c

```diff
@@ -0,0 +1,571 @@
+		goto err_slab_calloc;
+	}
+
+	total_size = handle->page_size * __page_count(handle);
+	GNIX_DEBUG(FI_LOG_EP_CTRL, "total_size requested for mmap: %zu.\n",
+		   total_size);
+
+	slab->used = calloc(1, sizeof(*(slab->used)));
+	if (!slab->used) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating bitmap: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err_bitmap_calloc;
+	}
+
+	slab->base = mmap(0, total_size, (PROT_READ | PROT_WRITE), MAP_SHARED,
+			  handle->fd, handle->last_offset);
+	if (slab->base == MAP_FAILED) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "%s\n", strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err_mmap;
+	}
+
+	ret = alloc_bitmap(slab->used, __mbox_count(handle));
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error allocating bitmap.\n");
+		goto err_alloc_bitmap;
+	}
+
+	status = GNI_MemRegister(handle->nic_handle->gni_nic_hndl,
+				 (uint64_t) slab->base, total_size,
+				 handle->cq_handle,
+				 GNI_MEM_READWRITE, -1,
+				 &slab->memory_handle);
+	if (status != GNI_RC_SUCCESS) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "GNI_MemRegister failed: %s\n",
+			  gni_err_str[status]);
+		ret = gnixu_to_fi_errno(status);
+		goto err_memregister;
+	}
+
+	slab->allocator = handle;
+
+	slist_insert_tail(&slab->list_entry, &handle->slab_list);
+
+	handle->last_offset += total_size;
+
+	return ret;
+
+err_memregister:
+	free_bitmap(slab->used);
+	free(slab->used);
+err_alloc_bitmap:
+	munmap(slab->base, total_size);
+err_mmap:
+	free(slab->used);
+err_bitmap_calloc:
+	free(slab);
+err_slab_calloc:
+	return ret;
+}
+
+/*
+ * TODO:
+ * - How to handle memory that is still out...
+ */
+static int __destroy_slab(struct gnix_mbox_alloc_handle *handle,
+			  struct gnix_slab *slab)
+{
+	size_t total_size;
+
+	if (!handle || !slab) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid argument handle or slab.\n");
+		return -FI_EINVAL;
+	}
+
+	total_size = handle->page_size * __page_count(handle);
+
+	free_bitmap(slab->used);
+	free(slab->used);
+
+	GNI_MemDeregister(handle->nic_handle->gni_nic_hndl,
+			  &slab->memory_handle);
+
+	munmap(slab->base, total_size);
+
+	free(slab);
+
+	return FI_SUCCESS;
+}
+
+static int __check_bitmap(struct gnix_mbox_alloc_handle *handle,
+			  struct gnix_slab **slab,
+			  int (*action)(gnix_bitmap_t *))
+{
+	struct slist_entry *entry;
+	struct gnix_slab *temp;
+	int ret;
+
+	*slab = NULL;
+
+	for (entry = handle->slab_list.head; entry; entry = entry->next) {
+		temp = container_of(entry, struct gnix_slab, list_entry);
+		ret = action(temp->used);
+		if (ret >= 0) {
+			*slab = temp;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static int __find_free(struct gnix_mbox_alloc_handle *handle,
+		       struct gnix_slab **slab)
+{
+	return __check_bitmap(handle, slab, find_first_zero_bit);
+}
+
+static int __find_used(struct gnix_mbox_alloc_handle *handle,
+		       struct gnix_slab **slab)
+{
+	return __check_bitmap(handle, slab, find_first_set_bit);
+}
+
+static int __fill_mbox(struct gnix_mbox_alloc_handle *handle,
```

Comment:
You will need to add comments to some of these functions as it isn't apparent what they do.